src/code/TwoHeadArena.c

```diff
@@ -1,140 +1,135 @@
-    return THGA_AllocEnd(thga, count * 0x40);
-}
-
-Gfx* THGA_AllocEnd64(TwoHeadGfxArena* thga) {
-    return THGA_AllocEnd(thga, 0x40);
-}
-
-Gfx* THGA_AllocEndArray16(TwoHeadGfxArena* thga, u32 count) {
-    return THGA_AllocEnd(thga, count * 0x10);
-}
-
-Gfx* THGA_AllocEnd16(TwoHeadGfxArena* thga) {
-    return THGA_AllocEnd(thga, 0x10);
-}
-
 void* THA_GetHead(TwoHeadArena* tha) {
     return tha->head;
 }
 
-void THA_SetHead(TwoHeadArena* tha, void* start) {
-    tha->head = start;
+void THA_SetHead(TwoHeadArena* tha, void* newHead) {
+    tha->head = newHead;
 }
 
 void* THA_GetTail(TwoHeadArena* tha) {
     return tha->tail;
 }
 
-void* THA_AllocStart(TwoHeadArena* tha, u32 size) {
+/**
+ * Allocates to the head of the Two Head Arena. The allocation will not have any alignment guarantees.
+ */
+void* THA_AllocHead(TwoHeadArena* tha, size_t size) {
     void* start = tha->head;
 
-    tha->head = (void*)((u32)tha->head + size);
+    tha->head = (u8*)tha->head + size;
     return start;
 }
 
-void* THA_AllocStart1(TwoHeadArena* tha) {
-    return THA_AllocStart(tha, 1);
+void* THA_AllocHead1(TwoHeadArena* tha) {
+    return THA_AllocHead(tha, 1);
 }
 
-void* THA_AllocEnd(TwoHeadArena* tha, u32 size) {
-    u32 mask;
+/**
+ * Allocates to the tail end of the Two Head Arena. The allocation will be aligned based on the size of the allocation.
+ * For all allocations greater than 16-bytes, they will be aligned to 16-bytes. Otherwise, the alignment will be the
+ * largest power of 2 for which the size is a multiple, in order to accommodate the alignment requirements of any data
+ * types that can fit within the allocation.
+ */
+void* THA_AllocTail(TwoHeadArena* tha, size_t size) {
+    uintptr_t mask;
 
     if (size == 8) {
-        mask = ~7;
+        // Align 8 for multiples of 8
+        mask = ALIGN_MASK(8);
     } else if (size == 4 || size == 12) {
-        mask = ~3;
+        // Align 4 for multiples of 4
+        mask = ALIGN_MASK(4);
     } else if (size == 2 || size == 6 || size == 10 || size == 12 || size == 14) {
-        mask = ~1;
+        // Align 2 for multiples of 2
+        mask = ALIGN_MASK(2);
+    } else if (size >= 0x10) {
+        // Align 0x10 for allocations greater than 0x10
+        mask = ALIGN_MASK(0x10);
     } else {
-        mask = (size >= 0x10) ? ~0xF : 0;
+        //! @bug if size is 1 the computation below will give NULL. The mask for this case would be
+        //! more sensible as ~0, for no extra alignment
+        mask = 0;
     }
 
-    tha->tail = (void*)((((u32)tha->tail & mask) - size) & mask);
+    tha->tail = (void*)((((uintptr_t)tha->tail & mask) - size) & mask);
     return tha->tail;
 }
 
-void* THA_AllocEndAlign16(TwoHeadArena* tha, u32 size) {
-    u32 mask = ~0xF;
+/**
+ * Allocates to the tail end of the Two Head Arena with guaranteed 16-byte alignment.
+ */
+void* THA_AllocTailAlign16(TwoHeadArena* tha, size_t size) {
+    uintptr_t mask = ALIGN_MASK(0x10);
```

Comment:
I find it weird to have 16 in the name and 0x10 in the code, but it's not a big deal and I understand why we either don't want 0x10 in the function name, or "hide" that 16 has a particular bit layout